prow/cmd/peribolos/main.go

```diff
@@ -870,3 +1013,330 @@ func configureTeamMembers(client teamMembersClient, id int, team org.Team, invit
+	}
+	if affiliation == string(github.AffiliationDirect) {
+		users = sets.NewString(users...).Intersection(members).List()
+	}
+	for _, user := range users {
+		permission, err := client.GetUserPermission(orgName, repo, user)
+		if err != nil {
+			return err
+		}
+		switch permission {
+		case string(org.Admin):
+			haveAdmin.Insert(user)
+		case string(org.Write):
+			haveWrite.Insert(user)
+		case string(org.Read):
+			haveRead.Insert(user)
+		}
+	}
+
+	have := permissionLevels{admin: haveAdmin, write: haveWrite, read: haveRead}
+	want := permissionLevels{admin: wantAdmin, write: wantWrite, read: wantRead}
+	have.normalize()
+	want.normalize()
+
+	// figure out who to remove
+	remove := have.all().Difference(want.all())
+
+	// sanity check changes
+	if d := float64(len(remove)) / float64(len(have.all())); d > opt.maximumDelta {
+		return fmt.Errorf("cannot delete %d repo collaborators or %.3f of %s/%s (exceeds limit of %.3f)", len(remove), d, orgName, repo, opt.maximumDelta)
+	}
+
+	adder := func(user, permission string) error {
+		if invitees.Has(user) { // Do not add them, as this causes another invite.
+			logrus.Infof("Waiting for %s to accept invitation as a collaborator to %s/%s", user, orgName, repo)
+			return nil
+		}
+
+		isMember := members.Has(user)
+		if isMember && affiliation == string(github.AffiliationOutside) {
+			logrus.Infof("Cannot add/update %s as outside collaborator to %s/%s since they are a member", user, orgName, repo)
+			return nil
+		}
+
+		if !isMember && affiliation == string(github.AffiliationDirect) {
+			logrus.Infof("%s is not a member of the org %s, retry as an outside collaborator or apply for membership", user, orgName)
+			return nil
+		}
+
+		err := client.UpdateCollaborator(orgName, repo, user, permission)
+		if err != nil {
+			logrus.WithError(err).Warnf("UpdateCollaborator(%s, %s, %s, %s) failed", orgName, repo, user, permission)
+		} else {
+			logrus.Infof("Set %s with %s access to %s/%s", user, permission, orgName, repo)
+		}
+		return err
+	}
+
+	remover := func(user string) error {
+		err := client.RemoveCollaborator(orgName, repo, user)
+		if err != nil {
+			logrus.WithError(err).Warnf("RemoveCollaborator(%s, %s, %s) failed", orgName, repo, user)
+		}
+		return err
+	}
+
+	return configureCollaborators(have, want, invitees, adder, remover)
+}
+
+func configureTeamCollaborators(opt options, client repoClient, orgName, repo string, config org.UserPermissionLevel, invitees sets.String) error {
+	// get desired state
+	wantAdmin := sets.NewString(config.Admin...)
+	wantWrite := sets.NewString(config.Write...)
+	wantRead := sets.NewString(config.Read...)
+
+	// get current state
+	teams, err := client.ListTeamsForRepo(orgName, repo)
+	if err != nil {
+		return err
+	}
+	var haveAdmin, haveWrite, haveRead sets.String
+	namesToID := map[string]int{}
+	for _, team := range teams {
+		namesToID[team.Name] = team.ID
+		switch team.Permission {
+		case string(org.Admin):
+			haveAdmin.Insert(team.Name)
```

Comment:
This has edge case failures, as team names may be duplicated and the end namesToID[team.Name] may or may not have this permission